sdks/python/apache_beam/io/gcp/spannerio_test.py

```diff
@@ -0,0 +1,182 @@
+try:
+  from google.cloud import spanner
+  from apache_beam.io.gcp.spannerio import ReadOperation # pylint: disable=unused-import
+  from apache_beam.io.gcp.spannerio import ReadFromSpanner # pylint: disable=unused-import
+except ImportError:
+  spanner = None
+# pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
+
+
+MAX_DB_NAME_LENGTH = 30
+TEST_PROJECT_ID = 'apache-beam-testing'
+TEST_INSTANCE_ID = 'beam-test'
+TEST_DATABASE_PREFIX = 'spanner-testdb-'
+# TEST_TABLE = 'users'
+# TEST_COLUMNS = ['Key', 'Value']
+FAKE_ROWS = [[1, 'Alice'], [2, 'Bob'], [3, 'Carl'], [4, 'Dan'], [5, 'Evan'],
+             [6, 'Floyd']]
+
+
+def _generate_database_name():
+  mask = string.ascii_lowercase + string.digits
+  length = MAX_DB_NAME_LENGTH - 1 - len(TEST_DATABASE_PREFIX)
+  return TEST_DATABASE_PREFIX + ''.join(random.choice(mask) for i in range(
+      length))
+
+
+def _generate_test_data():
+  mask = string.ascii_lowercase + string.digits
+  length = 100
+  return [('users', ['Key', 'Value'], [(x, ''.join(
+      random.choice(mask) for _ in range(length))) for x in range(1, 5)])]
+
+
+@unittest.skipIf(spanner is None, 'GCP dependencies are not installed.')
+class SpannerReadTest(unittest.TestCase):
+
+  @mock.patch('apache_beam.io.gcp.spannerio.Client')
+  @mock.patch('apache_beam.io.gcp.spannerio.BatchSnapshot')
+  def test_read_with_query_batch(self, mock_batch_snapshot_class,
+                                 mock_client_class):
+    mock_client = mock.MagicMock()
+    mock_instance = mock.MagicMock()
+    mock_database = mock.MagicMock()
+    mock_snapshot = mock.MagicMock()
+
+    mock_client_class.return_value = mock_client
+    mock_client.instance.return_value = mock_instance
+    mock_instance.database.return_value = mock_database
+    mock_database.batch_snapshot.return_value = mock_snapshot
+    mock_batch_snapshot_class.return_value = mock_snapshot
+    mock_batch_snapshot_class.from_dict.return_value = mock_snapshot
+
+    mock_snapshot.to_dict.return_value = dict()
+    mock_snapshot.generate_query_batches.return_value = [
+        {'query': {'sql': 'SELECT * FROM users'},
+         'partition': 'test_partition'} for _ in range(3)]
+    mock_snapshot.process_query_batch.side_effect = [
+        FAKE_ROWS[0:2], FAKE_ROWS[2:4], FAKE_ROWS[4:]]
+
+    pipeline = TestPipeline()
+    records = (pipeline
+               | ReadFromSpanner(
+                   project_id=TEST_PROJECT_ID,
+                   instance_id=TEST_INSTANCE_ID,
+                   database_id=_generate_database_name())
+               .with_query('SELECT * FROM users'))
+    pipeline.run()
+    assert_that(records, equal_to(FAKE_ROWS))
+
+  @mock.patch('apache_beam.io.gcp.spannerio.Client')
+  @mock.patch('apache_beam.io.gcp.spannerio.BatchSnapshot')
+  def test_read_with_table_batch(self, mock_batch_snapshot_class,
+                                 mock_client_class):
+    mock_client = mock.MagicMock()
```

Comment:
Also please consider adding an integration test similar to following for BigQuery.

https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_read_it_test.py

Reply:
Sure! I'll add some more tests as per the reference.

Reply:
I just saw the ticket (https://issues.apache.org/jira/browse/BEAM-7246), it says IT not included on this ticket.
Please suggest!